#2572 Added FreeBSD Operating System Support

endyman opened the pull request:
Added Support for FreeBSD installation. In order to install FreeBSD Servers using foreman you need to create a mfsbsd image using https://github.com/mmatuska/mfsbsd 

As we require rc.local support in order to start the sysinstall process i submitted a patch to mfsbsd, too https://github.com/mmatuska/mfsbsd/pull/17 

You can find a fork here: https://github.com/endyman/mfsbsd/tree/rc-local-support
Simply add a rc.local to /conf/ including the follwing code:

```
# create install.cfg for the installer
# if your forman server is not named foreman,
# change the name accordinly.

cat << EOF >/tmp/install.cfg
pc_config: https://foreman/unattended/provision
nic_config: dhcp-all
confirm_install: no
EOF

# start the installation process using pc-sysinstall

pc-sysinstall start-autoinstall /tmp/install.cfg
shutdown -r now
```

In case your foreman server is not called foreman, change it here. Currently there is no elegant way to pass params to BSD so we are using a static host here. I created a foreman CNAME in each domain pointing to the foreman server. After the kernel is booted, mfsbsd will try to fetch the installer provisioning config from https://foreman/unattended/provision

Next you need to build the mfsbsd image - BASE points to you install cdrom:

```
$ make BASE=/cdrom/usr/freebsd-dist
```

As a last step you need to copy the image on a webserve follwing this schema:

```
<www_dir>/boot/x86_64/images/mfsbsd.img
```

In foreman you simply point your install media to `http://<www_dir>/`

endyman committed:
#2572 Added FreeBSD Operating System Support

mmoll commented:
@endyman I was also thinking of adding support for PXE installing FreeBSD, but hadn't started yet :)
One of the problems to solve is the passing of parameters into the mfs-image. E.g. 

```
pc_config: https://foreman/unattended/provision
```

won't work for me as I'm using UUID tokens. But I have no idea how to pass this to the install environment running on the client...

endyman commented:
@mmoll Do you use UUID tokens for the network config? Everything else can be fetched prior install during mfs boot - that's how I implemented it and thats what the provision template is for.
I currently do not have a solution for passing args like preseed URL or nic configs therefore I implemented this kind of workaround. I decided that in most scenarios it will be sufficient to boot the install mfs via tftp/dhcp and let it fetch everything server specific from a known location afterwards (hardcoded in the mfs and resolvable via dns), before starting the installer. I think this is an acceptable trade-off as I still have no clue how to solve the arg passing issue in FreeBSD. If anyone reading this has an idea how to pass args to the loader and from the loader to the kernel and access such data post boot i.e. via kenv from user space please let me know!

ohadlevy commented on app/views/unattended/memdisk.rhtml:
```diff
@@ -0,0 +1,69 @@
+
+# install source configuration
+installMedium=ftp
+ftpPath=ftp://ftp2.de.freebsd.org/pub/FreeBSD/releases/amd64/9.1-RELEASE
```
shouldnt this be point to an install media in foreman?

endyman replied:
Currently the media and the mirror are two different things as freeBSD does not ship a useable install media within the repo like i.e. debian does. I use the foreman media to define the install bsd ramdisk image location. Otherwise I would have to introduce additional fields specially for bsd to specify the boot image location and a mirror location. But maybe the template can be improved by introducing architecture tokens.

ohadlevy commented:
[test]

ohadlevy commented:
looks pretty good start, would using ipxe instead of plain gpxe can solve the foreman path issue? is it possible to pass a parameter (like kernel args) in the template?

endyman commented:
I think ipxe will not help here. The problem is that mboot.c32 does not work anymore (which would allow kernel args). Memdisk in turn due to its generic approach and the loader in between  does not offer passing arguments to the kernel. Also using iPXE to natively boot a FreeBSD kernel AFAIK only works for i386. Unfortunately I think there are not to many options.

endyman commented:
While evaluating other options I came up with a more concise implementation: 

The medium_path will be used for the BSD mirror like for the other operating systems, too. The kernel is static as we use memdisk and the initrd will follow a scheme similar to Solaris i.e. FreeBSD-5.1-d-mfs.img. The Admin will have to take care that for each OS a mfs.img is available just like the admin has to care for the pxegrub on solaris.

This would also enable us to use the medium path in templates. I will create and test this approach and in case everything works well close this one and open a new PR.

endyman commented:
new PR: https://github.com/theforeman/foreman/pull/673
